internal/db/fetcher/indexer_iterators.go

```diff
@@ -689,22 +798,40 @@ func (f *IndexFetcher) determineFieldFilterConditions() ([]fieldFilterCond, erro
 
 			condMap := indexFilterCond.(map[connor.FilterKey]any)
 			for key, filterVal := range condMap {
-				opKey := key.(*mapper.Operator)
-				var normalVal client.NormalValue
+				cond := fieldFilterCond{
+					op:   key.(*mapper.Operator).Operation,
+					kind: f.indexedFields[i].Kind,
+				}
+
 				var err error
 				if filterVal == nil {
-					normalVal, err = client.NewNormalNil(f.indexedFields[i].Kind)
+					cond.val, err = client.NewNormalNil(cond.kind)
+				} else if !f.indexedFields[i].Kind.IsArray() {
+					cond.val, err = client.NewNormalValue(filterVal)
 				} else {
-					normalVal, err = client.NewNormalValue(filterVal)
+					subCondMap := filterVal.(map[connor.FilterKey]any)
+					for subKey, subVal := range subCondMap {
+						arrKind, ok := cond.kind.(client.ScalarArrayKind)
+						if !ok {
+							return nil, NewErrNotSupportedKindByIndex(cond.kind)
```

Comment:
suggestion: This line looks impossible to hit, I suggest deleting it and the error type and just do `arrKind := cond.kind.(client.ScalarArrayKind)`.  By including the check you suggest to readers that it is possible.  No worries if you feel strongly about keeping it though.

Reply:
I'm fine with deleting it. I even put into this place a TODO before, something like this: `TODO: can this be tests?` :)